src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,197 @@
+                    signalRegistrations = new List<WeakReference<PosixSignalRegistration>>();
+                    s_registrations.Add(_signo, signalRegistrations);
+                }
+                if (signalRegistrations.Count == 0)
+                {
+                    Interop.Sys.EnablePosixSignalHandling(_signo);
+                }
+                signalRegistrations.Add(new WeakReference<PosixSignalRegistration>(this));
+            }
+            _registered = true;
+        }
+
+        private void Handle(PosixSignalContext context)
+        {
+            lock (_gate)
+            {
+                if (_registered)
+                {
+                    _handler(context);
+                }
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static int OnPosixSignal(int signo, PosixSignal signal)
+        {
+            lock (s_registrations)
+            {
+                if (s_registrations.TryGetValue(signo, out List<WeakReference<PosixSignalRegistration>>? signalRegistrations))
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        var registrations = new PosixSignalRegistration?[signalRegistrations.Count];
+                        bool hasRegistrations = false;
+                        for (int i = 0; i < signalRegistrations.Count; i++)
+                        {
+                            if (signalRegistrations[i].TryGetTarget(out PosixSignalRegistration? registration))
+                            {
+                                registrations[i] = registration;
+                                hasRegistrations = true;
+                            }
+                        }
+                        if (hasRegistrations)
+                        {
+                            // This is called on the native signal handling thread. We need to move to another thread so
+                            // signal handling is not blocked. Otherwise we may get deadlocked when the handler depends
+                            // on work triggered from the signal handling thread.
+
+                            // For terminate/interrupt signals we use a dedicated Thread
+                            // in case the ThreadPool is saturated.
+                            bool useDedicatedThread = signal == PosixSignal.SIGINT ||
+                                                      signal == PosixSignal.SIGQUIT ||
+                                                      signal == PosixSignal.SIGTERM;
+                            if (useDedicatedThread)
+                            {
+                                Thread handlerThread = new Thread(HandleSignal)
+                                {
+                                    IsBackground = true,
+                                    Name = ".NET Signal Handler"
+                                };
+                                handlerThread.Start((signo, registrations));
```

Comment:
```suggestion
                                handlerThread.UnsafeStart((signo, registrations));
```
to match `UnsafeQueueUserWorkItem` in else